Update ufs model to version from 10-12-23

JessicaMeixner-NOAA opened the pull request:
# Description

This updates the ufs-weather-model to the commit hash from 10-12-23 from the HR2 tag.  The diffs can be seen here: https://github.com/ufs-community/ufs-weather-model/compare/GFSv17.HR2...68050e58589a82ab509aaefaafdc2a6b90f34e48  This replaces PR https://github.com/NOAA-EMC/global-workflow/pull/1889 

Resolves https://github.com/NOAA-EMC/global-workflow/issues/1811


# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.) - routine update of model 

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Free forecast for low resolution S2SWA on hera (previously tested other machines as well). 
-  This is ready for additional testing 

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

JessicaMeixner-NOAA committed:
update model hash to latest

JessicaMeixner-NOAA committed:
update configure files based on model updates to reg test

JessicaMeixner-NOAA committed:
updates to wcoss2 for module conflict errors

JessicaMeixner-NOAA committed:
temporary updates to not link files that caused issues with aerosols

JessicaMeixner-NOAA committed:
update model version (temporary branch) to use stack 1.5
and update for prod_util/prod-util (temporary diff)

JessicaMeixner-NOAA committed:
update for gocart new updates

JessicaMeixner-NOAA committed:
revert back to ufs develop branch

JessicaMeixner-NOAA committed:
update wave jobs to new modules

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/updateufsstack15

Conflicts:
	ush/nems_configure.sh

JessicaMeixner-NOAA committed:
add coupled script temporarily

JessicaMeixner-NOAA committed:
revert wcoss2 updates finish backing out 1.5 updates and just got to top
of develop

JessicaMeixner-NOAA committed:
add wgrib2

JessicaMeixner-NOAA committed:
Merge branch 'feature/updateufsstack15' of github.com:JessicaMeixner-NOAA/global-workflow into feature/updateufsstack15

JessicaMeixner-NOAA committed:
update modules for wcoss2

JessicaMeixner-NOAA committed:
update for init cfp job on wcoss2

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/updateufsstack15

JessicaMeixner-NOAA committed:
use latest UFS

JessicaMeixner-NOAA committed:
update fcst and waveinit modules

JessicaMeixner-NOAA committed:
load extra module for wcoss2 wave jobs in rocoto script

JessicaMeixner-NOAA committed:
update UFS to latest

JessicaMeixner-NOAA committed:
move linked gocart files to a copy statement at the end of run
due to crashes in gocart

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into updateUFS101223

JessicaMeixner-NOAA committed:
update copying of gocart files

JessicaMeixner-NOAA committed:
debugged output for gocart

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into updateUFS101223

JessicaMeixner-NOAA committed:
remove temporary file

JessicaMeixner-NOAA committed:
Merge branch 'updateUFS101223' of github.com:JessicaMeixner-NOAA/global-workflow into updateUFS101223

JessicaMeixner-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -1038,7 +1038,24 @@
+  # TO DO: this should be linked but there were issues where gocart was crashing if it was linked
+  #${NCP} "${DATA}/gocart.inst_aod.*" "${COM_CHEM_HISTORY}/"
+  if (( $( ls -1 "${DATA}/gocart.inst_aod."* 2> /dev/null | wc -l) > 0 )); then
+    for file in $(ls "${DATA}/gocart.inst_aod."*) ; do
```
@WalterKolczynski-NOAA --- I've tried a lot of things to get the gocart output back to COM instead of it being linked which crashed the model.  This was literally the only way I can get this to work.  I also copied this from another place in the code.  If theres code someone would prefer instead of this, I'm happy to try it, but I need help if this is going to be a blocking issue for this PR to get through.  I've already spent a week + working on this and I have no idea what to try if this isn't going to be acceptable.

DavidHuber-NOAA replied:
You may be able to use [this suggestion from ShellCheck](https://www.shellcheck.net/wiki/SC2207):

```
declare -a files
while IFS='' read -r line; do files+=("$line"); done < <( ls "${DATA}/gocart.inst_aod."* 2> /dev/null )
for file in "${files[@]}"; do
   ${NCP} "${file}"  "${COM_CHEM_HISTORY}/"
done
```

WalterKolczynski-NOAA replied:
I'm testing something better

JessicaMeixner-NOAA replied:
@DavidHuber-NOAA  thank you for this suggestion! I was just getting a chance to test this suggestion and I saw @WalterKolczynski-NOAA comment, so I'll hold off on testing that until I hear more from Walter.  Thanks Walter for taking on updating this! Please let m know if there's anything I can do from my end.

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -1038,7 +1038,24 @@ GOCART_postdet() {
+  # Copy gocart.inst_aod after the forecast is run (and successfull)
+  # TO DO: this should be linked but there were issues where gocart was crashing if it was linked
+  #${NCP} "${DATA}/gocart.inst_aod.*" "${COM_CHEM_HISTORY}/"
+  if (( $( ls -1 "${DATA}/gocart.inst_aod."* 2> /dev/null | wc -l) > 0 )); then
+    for file in $(ls "${DATA}/gocart.inst_aod."*) ; do
+      ${NCP} "${file}"  "${COM_CHEM_HISTORY}/"
+    done
+  fi
```
This works for me on WCOSS:
```suggestion
  for fhr in ${FV3_OUTPUT_FH}; do
    if (( fhr == 0 )); then continue; fi
    local vdate=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${fhr} hours" +%Y%m%d%H)
    ${NCP} "${DATA}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4" \
      "${COM_CHEM_HISTORY}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4"
  done
```

Note: I originally tried to test on Orion and the forecast failed pretty quickly. I think there may be something still wrong there.

JessicaMeixner-NOAA replied:
I'll try to run the branch as in on orion before implementing this change.   I don't remember issues running on orion and there were stmp issues overnight so hopefully it's simply was a space issue.

JessicaMeixner-NOAA committed:
Update ush/forecast_postdet.sh

Suggested update from Walter

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA and @DavidHuber-NOAA thank you for your help addressing the shel check issue!   I have committed Walter's recommendation.  My forecast test on orion succeeded: /work2/noaa/marine/jmeixner/HR3/updatemodel/s2swc48t01/COMROOT/s2swc48t01/logs/2021032312/gfsfcst.log I suspect @WalterKolczynski-NOAA issue was an stmp space issue, but happy take a closer look.

JessicaMeixner-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -1038,7 +1038,25 @@
+  #${NCP} "${DATA}/gocart.inst_aod.*" "${COM_CHEM_HISTORY}/"
+  for fhr in ${FV3_OUTPUT_FH}; do
+    if (( fhr == 0 )); then continue; fi
+    local vdate=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${fhr} hours" +%Y%m%d%H)
```
@WalterKolczynski-NOAA  Thoughts on how to address this new warning or if it can be ignored?

aerorahul replied:
one could do the following:
before the `for` loop, declare `vdate` as `local` as:
```
local vdate
```
And then on this line remove the declaration `local`. 
Just a suggestion.

aerorahul replied:
`fhr` should also be a local variable; technically speaking.

JessicaMeixner-NOAA committed:
implementing review suggestions

JessicaMeixner-NOAA commented:
Thanks @aerorahul  for the suggestions.  This now passes shell check.  Please let me know if any re-testing is now required on my end.

aerorahul commented on jobs/rocoto/wavepostbndpnt.sh:
```diff
@@ -5,9 +5,65 @@ source "${HOMEgfs}/ush/preamble.sh"
+  else
+    module load prod-util
+  fi
+fi
+
+
+if [[ "${MACHINE_ID}" = "hera" ]]; then
+  module use "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/modulefiles/core"
+  module load "miniconda3/4.6.14"
+  module load "gfs_workflow/1.0.0"
+# TODO: orion and wcoss2 will be uncommented when they are ready.  This comment block will be removed in the next PR
+#elif [[ "${MACHINE_ID}" = "orion" ]]; then
+#  module use "/home/rmahajan/opt/global-workflow/modulefiles/core"
+#  module load "python/3.7.5"
+#  module load "gfs_workflow/1.0.0"
+#elif [[ "${MACHINE_ID}" = "wcoss2" ]]; then
+#  module load "python/3.7.5"
+fi
+if [[ "${MACHINE_ID}" == "noaacloud" ]]; then
+   if [[ "${PW_CSP:-}" = "aws" ]]; then
+
+      # TODO: This can be cleaned-up; most of this is a hack for now.
+      module use "/contrib/spack-stack/envs/ufswm/install/modulefiles/Core"
+      module load "stack-intel"
+      module load "stack-intel-oneapi-mpi"
+      module use -a "/contrib/spack-stack/miniconda/modulefiles/miniconda/"
+      module load "py39_4.12.0"
+      module load "ufs-weather-model-env/1.0.0"
+      export NETCDF="/contrib/spack-stack/miniconda/apps/miniconda/py39_4.12.0"
+      # TODO: Are there plans for EPIC to maintain this package or should GW provide support?
+      export UTILROOT="/contrib/global-workflow/NCEPLIBS-prod_util"
+      export PATH="${PATH}:/contrib/global-workflow/bin"
+      ndate_path="$(command -v ndate)"
+      export NDATE="${ndate_path}"
+   fi
+fi
+
+module list
+unset MACHINE_ID
+set_trace
```
Can I make a recommendation of placing this in a separate script e.g. `ush/load_ufswm_modules.sh` and sourcing that here and in the other wave, fcst and efcs scripts instead of repeating this block everywhere?

JessicaMeixner-NOAA replied:
This is a great suggestion however some of the jobs require additional module loads that others don't and I don't know if you want the extra modules for all the jobs or not?  I guess the script could just be for the default and then added ones to the specific modules or we could have all jobs load the extra modules....   Thoughts? 

One other question I have before implementing this, it is my understanding is that this is a quick temporary fix and not expected to stay long term.  Do we want to go through the effort if it's not long term?

aerorahul replied:
IMO, it is worth the effort even if it is temporary, because we don't know how long will it take.  It also eliminates the possibility of copy/paste error.  In addition, it makes it clear which modules are being loaded and from where.

The `export USE_CFP=YES` line the `waveinit.sh` script should be in the section of `waveinit` in `env/WCOSS2.env`.
That just leaves loading the `cfp` module for wcoss2 on that job.  That can be loaded in the job in question.  It clearly identifies which modules are needed in addition to those in the ufs-weather-model.

JessicaMeixner-NOAA replied:
Thanks @aerorahul  -  I will implement these updates today.

JessicaMeixner-NOAA replied:
Updates are here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/updateUFS101223module will test before merging to this branch. Test has started on orion, building on hera and wcoss2... should know more this afternoon.

JessicaMeixner-NOAA replied:
Updates have currently been tested on orion and wcoss2.   WCOSS2 failed in verif and I've seen others have failures in this job recently, so it's unclear if its related to this PR (i'm guessing not).  Otherwise things are fine, but the point job is taking a  longer so we either need to decrease the frequency of the point output for testing and/or increase the wall time.     Testing on hera is about to get started, but not quite sure what to do about the vrfy job though, I guess I'm going to test develop next?  Open to other ideas if anyone has them.

JessicaMeixner-NOAA committed:
updates to clean up module load and create function to reduce possibility of error

JessicaMeixner-NOAA committed:
update pnt resources and remove temp file

JessicaMeixner-NOAA commented:
Module loads have been cleaned up and improved following suggestion of @aerorahul 

The wave point job is taking longer than before so some time was added to that job. 

The wcoss2 test did have an issue, but that issue is also in develop https://github.com/NOAA-EMC/global-workflow/issues/1947 and unfortunately I'm not really sure how to solve that problem

JessicaMeixner-NOAA commented:
The shell-check warnings are on lines of code that I did not change, so I believe that means I do not need to address that.  If that's no longer the case, please let me know how this should be updated.

WalterKolczynski-NOAA commented on jobs/rocoto/waveprep.sh:
```diff
@@ -5,10 +5,12 @@ source "${HOMEgfs}/ush/preamble.sh"
 status=$?
 [[ ${status} -ne 0 ]] && exit ${status}
 
+
```
```suggestion
```

WalterKolczynski-NOAA commented on ush/load_ufswm_modules.sh:
```diff
@@ -0,0 +1,71 @@
+ulimit_s=$( ulimit -S -s )
+
+source "${HOMEgfs}/ush/detect_machine.sh"
+set +x
```
This is already being done on L6
```suggestion
```

WalterKolczynski-NOAA commented on ush/load_ufswm_modules.sh:
```diff
@@ -0,0 +1,71 @@
+    module load cfp
+    module load libjpeg
+  else
+     module load prod-util
```
```suggestion
    module load prod-util
```

WalterKolczynski-NOAA commented on Externals.cfg:
```diff
@@ -8,7 +8,7 @@ protocol = git
 
 [UFS]
-tag = 4d05445
+tag = 68050e5 
```
```suggestion
tag = 68050e5
```

WalterKolczynski-NOAA reviewed (changes requested):
Very minor style stuff. Sending it through CI in the meantime.

WalterKolczynski-NOAA commented:
GEFS test is failing because `$NETCDF` is not being set. The GFS version of `config.base` has a default empty string if it isn't set, which is why Jessica's self-tests ran, but the GEFS version does not.

- This suggests the module being loaded by UFS is not setting the variable but the module workflow is loading does (a recurring issue when changing module version)
  - Should the module be setting this? If so, we need to contact either the EPIC or NCEPLibs team.
- Short-term, @JessicaMeixner-NOAA please just change the `${NETCDF}` in https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/config/gefs/config.base.emc.dyn#L85 to `${NETCDF:-}`. That will hopefully allow these tests to pass and the PR to be merged.
- Long-term, (@KateFriedman-NOAA @aerorahul) I don't think we really should be setting this in `config.base` at all. I think only select obs is using this value. A grep found these:
```
>grep NCDUMP -R
parm/config/gfs/config.base.emc.dyn:export NCDUMP="${NETCDF:-}/bin/ncdump"
parm/config/gefs/config.base.emc.dyn:export NCDUMP="${NETCDF}/bin/ncdump"
sorc/verif-global.fd/parm/metplus_config/machine.conf:NCDUMP = {ENV[NCDUMP]}
sorc/verif-global.fd/ush/run_verif_global_in_global_workflow.sh:    export NCDUMP=$(which ncdump)
sorc/verif-global.fd/ush/run_verif_global_in_global_workflow.sh:    export NCDUMP=$(which ncdump | sed 's/ncdump is //g')
sorc/verif-global.fd/ush/get_data_files.py:    ncdump = os.environ['NCDUMP']
sorc/verif-global.fd/ush/load_modules.sh:    export NCDUMP=$(which ncdump)
sorc/verif-global.fd/ush/load_modules.sh:    export NCDUMP=$(which ncdump | sed 's/ncdump is //g')
sorc/verif-global.fd/ush/create_METplus_job_scripts.py:        'CUT', 'TR', 'NCAP2', 'CONVERT', 'NCDUMP'
sorc/ufs_utils.fd/ush/fv3gfs_make_grid.sh:  res=$( $NCDUMP -h $1 | grep -o ":RES_equiv = [0-9]\+" | grep -o "[0-9]" )
sorc/ufs_utils.fd/ush/fv3gfs_driver_grid.sh:export NCDUMP=${NCDUMP:-ncdump}
sorc/ufs_utils.fd/ush/fv3gfs_driver_grid.sh:  res=$( $NCDUMP -h ${grid_dir}/C*_grid.tile7.nc | grep -o ":RES_equiv = [0-9]\+" | grep -o "[0-9]" )
jobs/JGDAS_ENKF_SELECT_OBS:LEVS=$(${NCDUMP} -h "${ATMGES_ENSMEAN}" | grep -i "pfull" | head -1 | awk -F" = " '{print $2}' | awk -F" " '{print $1}') # get LEVS
```
The ufs_utils ones I think are only offline tools. I haven't traced the path through the scripts, but it appears verif-global is setting its own value with `which` first. That leaves only the last entry, select obs. (This clone did not have GDAS App checked out, so possibly it is used there.)

JessicaMeixner-NOAA committed:
Update jobs/rocoto/waveprep.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

JessicaMeixner-NOAA committed:
Update ush/load_ufswm_modules.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

JessicaMeixner-NOAA committed:
Update Externals.cfg

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

JessicaMeixner-NOAA committed:
Update ush/load_ufswm_modules.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

JessicaMeixner-NOAA committed:
updating NETCDF as suggested by Walter for GEFS test

JessicaMeixner-NOAA committed:
Merge branch 'updateUFS101223' of github.com:JessicaMeixner-NOAA/global-workflow into updateUFS101223

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA thanks for your detailed review.  All of your minor suggestions have been committed and I just updated  ${NETCDF} in https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/config/gefs/config.base.emc.dyn#L85

Please let me know if there's anything else that I need to do from my end and I saw the comments on the verif issue as well.  This seemed a little odd to me because I thought that we had moved to the new COM in HR2 but I didn't remember seeing this error.

WalterKolczynski-NOAA commented:
Same problem, this time with `$UTILROOT`. `load_ufswm_modules.sh` is trying to load a `prod-util` instead of `prod_util` and it is showing in the module list, but I can't seem to find that module anywhere to examine it, only `prod_util`. Suggest changing L23 of `load_ufswm_modules.sh` to use the underscore version unless someone can point me where the hyphenated version actually is.

WalterKolczynski-NOAA commented:
Actually, hold on. I might have more luck finding it if I actually check out this branch.

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA thanks! Let me know if there's something I can do from my end.

WalterKolczynski-NOAA commented:
Okay, after some temporary struggles, I was able to get to the module. EPIC has decided to change the name of the variable from `UTILROOT` to `prod_util_ROOT`. Of course, this is a problem for us because the NCO version on WCOSS still uses `UTILROOT`.

@JessicaMeixner-NOAA For now, to get this through, add this to the `load_ufswm_modules.sh` after it is loaded on non-WCOSS2 HPC:
```
export UTILROOT=${prod_util_ROOT}
```

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA  thank you so much for your help.  The spack-stack team has been helpful in getting certain things in sync with what's on WCOSS2, for example prod-util will be prod_util in a future spack-stack release.  Sounds like we need to report this issue as well.  Is this something I need to do or has this already been reported?  In the meantime, thanks for finding a temporary work around.  I'll get this added now.

JessicaMeixner-NOAA committed:
add export UTILROOT=${prod_util_ROOT}

TerrenceMcGuinness-NOAA commented:
Not sure why the CI driver didn't didn't clone the repo on Hera (might be space).  I'm going to reset the Hera CI Run so I can look at the cron logs.

WalterKolczynski-NOAA commented:
@DavidHuber-NOAA you should be aware of the module issues we're hitting here (or maybe you've already run into them) as you move workflow to spack-stack.

WalterKolczynski-NOAA commented:
`jobs/rocoto/efcs.sh` needs to be updated in the same manner as `fcst.sh`

JessicaMeixner-NOAA committed:
update efcs in the same way fcst job was updated

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into updateUFS101223

JessicaMeixner-NOAA commented:
> `jobs/rocoto/efcs.sh` needs to be updated in the same manner as `fcst.sh`

Done - I also merged with develop.

JessicaMeixner-NOAA commented:
The point job failed due to wallclock time: 
slurmstepd: error: *** JOB 15528189 ON Orion-03-15 CANCELLED AT 2023-10-24T14:50:57 DUE TO TIME LIMIT ***

I've raised the limit, although I guess not enough.  Do you want to try re-running this or for me to raise the limit again

WalterKolczynski-NOAA commented:
@JessicaMeixner-NOAA Go ahead and make the change, since it will need to be committed anyway, but I am checking with @TerrenceMcGuinness-NOAA whether we can just rewind that job so we don't have to start from scratch again.

JessicaMeixner-NOAA committed:
pad point job with extra time to avoid out of time errors

JessicaMeixner-NOAA commented:
I did a pretty big jump to just avoid this. We have a long term project to redo this job  which is underway.

TerrenceMcGuinness-NOAA commented:
ok I can get C48_S2SW going on Orion again, just getting to that now

JessicaMeixner-NOAA commented:
Thanks @TerrenceMcGuinness-NOAA

TerrenceMcGuinness-NOAA commented:
Sure thing, I increased the wall clock time on **gfswavepostpnt** from 2 hours to 4 hours and did a **rocotorewind** on it and updated the label

```
(gfs_workflow) mterry (Orion-login-1) C48_S2SW_2db41ab9 $ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          15530380     orion C48_S2SW   mterry  R       0:08      5 Orion-07-[51-52],Orion-23-[20,32],Orion-25-40
```

Running log:  `/work2/noaa/stmp/GFS_CI_ROOT/PR/1933/RUNTESTS/COMROT/C48_S2SW_2db41ab9/logs/2021032312/gfswavepostpnt.log`

I'm sorry hacking the process didn't work out.  It will have to be restarted from scratch on Orion.

I restarted the CI run on Orion from scratch.  I will keep an eye on it and increase the wall clock time on **gfswavepostpnt** before it runs.

TerrenceMcGuinness-NOAA commented:
Oh yes I see on the rebuild the wall time for **fgswavepostpnt** was already set to 4 hours directly in the PR.  It's running now just fine with those settings in place as specified.

WalterKolczynski-NOAA commented:
@JessicaMeixner-NOAA I think this can be merged as soon as the conflicts are resolved.

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into updateUFS101223

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA  - merged conflicts fixed!
